app/controllers/kudos_controller.rb

```diff
@@ -57,6 +57,14 @@ def create
         end
       end
     end
+  rescue ActiveRecord::RecordNotUnique
+    # Uniqueness checks at application level (Rails validations) are inherently
+    # prone to race conditions. If we pass Rails validations but get rejected
+    # by database unique indices, use the usual duplicate error message.
+    #
+    # https://api.rubyonrails.org/v5.1/classes/ActiveRecord/Validations/ClassMethods.html#method-i-validates_uniqueness_of-label-Concurrency+and+integrity
+    flash[:comment_error] = ts("You have already left kudos here. :)")
+    redirect_to request.referer
```

Comment:
I'm not 100% certain, but I think not using a `format` block with separate code for `js` means that both the error and the redirect will get eaten when the user has javascript enabled (which is probably most of the time).

On the other hand, I think this error is most likely to occur if the user double-submits the form, in which case the javascript response returned by the first form submission would update the page, and the second error message would be unnecessary. So it might be good that the error and redirect will be eaten? IDK.

Reply:
Thanks! The redirect will be lost, I forgot to check for `format: :js`. This should be lawful-good-fixed now.